org.eclipse.scanning.api/src/org/eclipse/scanning/api/scan/ScanEstimator.java

```diff
@@ -97,35 +108,39 @@ public ScanEstimator(IPointGeneratorService pservice, ScanRequest<?> request, lo
 	 * @throws GeneratorException
 	 */
 	public ScanEstimator(IPointGenerator<?> gen, Map<String, Object> detectors, long timePerPoint) throws GeneratorException {
-
-		
-		this.generator = gen;
-		
+		this(gen, detectors.values(), timePerPoint);
+	}
+	
+	public ScanEstimator(Iterable<IPosition> positionIterable, Collection<Object> detectorModels, long timePerPoint) throws GeneratorException {
 		// TODO FIXME If some detectors are malcolm, they may have a wait time.
 		// If some are malcolm we may wish to ignore the input point time from the user
 		// in favour of the malcolm time per point or maybe the device tells us how long it will take?
-		if (detectors!=null) for (Object model : detectors.values()) {
-			if (model instanceof IDetectorModel) {
-				timePerPoint = Math.max(timePerPoint, Math.round(1000*((IDetectorModel)model).getExposureTime()));
-			}
+		if (detectorModels != null && !detectorModels.isEmpty()) {
+			timePerPoint = detectorModels.stream().filter(IDetectorModel.class::isInstance)
+					.map(m -> Math.round(((IDetectorModel) m).getExposureTime() * 1000))
+					.reduce(0l, Math::max);
 		}
-		this.size          = getEstimatedSize(gen);
-		this.rank          = gen.iterator().next().getScanRank(); // The rank of the scan is constant
-		this.timePerPoint  = timePerPoint;
-		this.estimatedScanTime      = size*timePerPoint;
+		
+		this.generator = positionIterable;
+		this.size = getEstimatedSize(positionIterable);
+		this.rank = positionIterable.iterator().next().getScanRank();
+		this.timePerPoint = timePerPoint;
+		this.estimatedScanTime = size * timePerPoint;
 	}
 
-	private int getEstimatedSize(IPointGenerator<?> gen) throws GeneratorException {
+	private int getEstimatedSize(Iterable<IPosition> gen) throws GeneratorException {
 		
 		int size=0;
 		if (gen instanceof IDeviceDependentIterable) {
 			size = ((IDeviceDependentIterable)gen).size();
 			
-		} else {
+		} else if (gen instanceof IPointGenerator) {
 			size = ((IPointGenerator<?>)gen).size();
-			
-		} 
-		return size;   		
+		} else if (gen!=null) {
+			for (IPosition unused : gen) size++; // Fast even for large stuff providing they do not check hardware on the next() call.
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused "unused" local variable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1481)